src/ol/loadImage.js

```diff
@@ -0,0 +1,46 @@
+
```

Comment:
For the purpose of a single occurrence, this is quite a lot of code. Couldn't you solve this by shimming `Image` in your worker? Something like
```js
 class ImageShim {
  constructor(opt_width, opt_height) {
    this.width = opt_width;
    this.height = opt_height;
  }
  set src(value) {
    // implement worker image loading here
  }
}

self.Image = ImageShim;
```

Reply:
The idea is very appealing and it would work well for instantiating an image-like object.
However consumption would not work: how could the shim be considered like a normal image by the browser? What API to use to return the data?

@ahocevar, I see there are 2 other places where we do `new Image()` in the code, if you are OK with introducing a `loadImage` utility function I can refactor to use it there too. What do you think?

Reply:
@gberaudo You would only shim this in a worker. And if your ImageShim is a `bitmaprenderer` canvas (instead of a plain object), with a src setter added, you should be fine.
```js
class ImageShim extends OffscreeenCanvas {
  set src(value) {
    // implement worker image loading here
  }
}
```
I'd hold off on using a `loadImage` function everywhere, at least until @fredj's experiments with `Image#decode` are completed in #9130.

Reply:
@ahocevar, I implemented your idea and it gives great results but is not compatible with transpilation to ES5. Here is the error message I get when extending the `OffscreenCanvas`:

`Please use the 'new' operator, this DOM object constructor cannot be called as a function`.

So, are you OK to go with my initial proposal?

Reply:
@gberaudo Please elaborate. Where exactly do you get this error? Which command sequence? I'm still not ok with your initial proposal, for the reasons already pointed out.

Reply:
@ahocevar, in https://github.com/openlayers/openlayers/commit/824cd16644232e07908f159f75a453a96c77bac0#diff-64dc856a6008c886102c4a48e74cbd6aR73, the transpiler transforms:

```js
export default class WorkerImage extends OffscreenCanvas {

  /**
   * @param {number} [width] Canvas width.
   * @param {number} [height] Canvas height.
   */
  constructor(width, height) {
    super(width || 0, height || 0);
    this.width_ = width;
    this.height_ = height;
    this.previousListeners_ = {};
  }
```

to

```js
var WorkerImage = /*@__PURE__*/(function (OffscreenCanvas) {
  function WorkerImage(width, height) {
    OffscreenCanvas.call(this, width || 0, height || 0);
    this.width_ = width;
    this.height_ = height;
    this.previousListeners_ = {};
  }

  if ( OffscreenCanvas ) WorkerImage.__proto__ = OffscreenCanvas;
  WorkerImage.prototype = Object.create( OffscreenCanvas && OffscreenCanvas.prototype );
  WorkerImage.prototype.constructor = WorkerImage;
```

The browser throws the above error when the constructor is called as a plain method: `OffscreenCanvas.call(this, width || 0, height || 0);`. That is why it works with ES6 code but not when transpiled to ES5. See https://github.com/gberaudo/openlayers/tree/webworker_poc_image_shim, note that transpilation is disabled in that branch.

Regarding your initial concerns about:
- the amount of code, I have refactored the code and most of it is in the example now;
- the nested function calls does not look like a performance issue (I ran tests in chrome and Safari) but I can look more into it if you think it is an issue.

Let me know what you think about all this. I am comited to find a suitable solution.

Reply:
Regarding Image shim I sent a [pull request](https://github.com/gberaudo/openlayers/pull/2) against your `webworker_poc_image_shim` branch.

Regarding try/catch, I meant that the try/catch should not be in a image loader that is used in places where no try/catch is needed. But I want you to remove the image load helper anyway, so this won't be an issue in the final pull request.

So when you switch this pull request to your `webworker_poc_image_shim` branch with my change merged, and maybe further simplify the image shim, I'll do a quick last review round and we should be good.

Reply:
OK @ahocevar, lets do like you propose: I merged your PR and will continue from there.